### PR TITLE
Implement connections list screen

### PR DESCRIPTION
**Pull request by dfalbel**
Adresses https://github.com/posit-dev/positron/issues/5107.

Demo usage:

https://github.com/user-attachments/assets/cd55ac78-01fd-4e7f-acc2-c60c7b86404f

<details>
<summary>
TODO:
</summary>

- [x] Rename action bar for specific schema navigation component
- [x] Initial implementation of the list connections view
- [x] Implement styles for the list connections view
- [x] Correctly design the action bar for the list view
- [x] Allow deleting a saved connection
- [x] Show only the selected connection in the schema view
- [x] Secondary action bar in the schema view (with details about the selected connection instance)
- [x] Don't show top level connection when exploring schema. Requires instances to always be in the expanded state - Ended up refactoring the cache structure.
- [x] Connection Instances should send `onChangeConnections` event when their status changes.
- [x] Resume connection modal
- [x] Add Copy button behavior for the resume connection modal
- [x] Add Edit button behavior for the connections modal
- [x] Add Resume connections button behavior
- [x] Refresh schema navigation action bar
- [x] Bug: Refreshing causes the expanded state to be reset.
- [x] Bug: Code is not being correctly restored when the IDE restarts.
- [x] Fix heights for the `List` component - so it scrolls correctly in the presence of multiple connections.
- [x] Make sure scrolluing works, and that we save scroll state
- [x] `%connection_view conn` should direct to schema explorer view
- [x] Remove filters for now
- [x] localize button titles in action bars

</details>

Follow up PR:

- [ ] Improve the display of code in the Resume connections modal - turns out it's not that simple to plug a text editor within a modal: https://github.com/posit-dev/positron/issues/5219

### CODE REVIEW

**Comment by petetronic** [comment]
Let's wait until we branch for 2024.11, and once main is open for prerelease 2024.12 changes we can review and plan to merging this on main (likely tomorrow).

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+					{ItemEntry}
+				</List>
+			</div>
+		</div>
+	);
+};
+
+interface ItemEntryProps {
+	index: number;
+	style: any;
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:145]: Better as `style: CSSProperties | undefined;` (or without the `undefined`, if style is required).

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+interface PositronConnectionsItemProps {
+	item: IPositronConnectionEntry;
+	style?: any;
+	selected: boolean;
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:151]: Better as `style?: CSSProperties;`

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+.connections-dtype {
+	color: var(--vscode-positronSideActionBar-disabledForeground);
+	font-weight: 300;
+}
+
+.connections-dtype::before {
+	content: ":";
+	color: var(--vscode-positronVariables-foreground);
+	margin-left: 3px;
+	margin-right: 3px;
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css:38]: NIT: `margin: 0 3px;`

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	color: var(--vscode-positronVariables-foreground);
+	margin-left: 0px;
+	margin-right: 6px;
+}
+
+.connections-item .connections-icon {
+	flex-shrink: 0;
+	flex-grow: 0;
+	padding-left: 5px;
+	padding-right: 5px;
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css:52]: NIT: padding: 0 5px;

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+			setEntries(entries);
+		}));
+
+		disposableStore.add(activeInstance.onDidChangeStatus((active) => {
+			if (!active) {
+				setActiveInstanceId(undefined);
+			}
+		}));
+
+		activeInstance.refreshEntries();
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:68]: `activeInstance.refreshEntries()` is not awaited / error handled.

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+							className={`codicon codicon-chevron-${props.item.expanded ? 'down' : 'right'}`}
+						>
+						</div>
+					</div>
+			}
+			<div
+				className='connections-details'
+				onMouseDown={rowMouseDownHandler}
+			>
+				<span className='connections-name'>{props.item.name}</span>
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:258]: I would write:

```
{props.item.dtype && <span className='connections-dtype'>{props.item.dtype}</span>}
{props.item.error && <span className='connections-error codicon codicon-error' title={props.item.error}></span>}
```

For theses two `span`s below:

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx [hunk 7]
```diff
@@ -73,38 +55,24 @@ export const ActionBar = (props: React.PropsWithChildren<ConnectionActionBarProp
 					<ActionBarRegion location='left'>
 						<ActionBarButton
 							align='left'
-							// TODO: should have a connect-icon
-							iconId='positron-new-connection'
-							tooltip={() => 'Connect'}
-							disabled={connectDisabled}
-							onPressed={() => props.selectedEntry?.connect?.()}
+							iconId='arrow-left'
+							tooltip={() => localize('positron.schemaNavigationActionBar.back', 'Back')}
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx:59]: Should be:

```
tooltip={(() => localize('positron.schemaNavigationActionBar.back', 'Back'))()}
```

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx [hunk 8]
```diff
@@ -73,38 +55,24 @@ export const ActionBar = (props: React.PropsWithChildren<ConnectionActionBarProp
...
+							onPressed={() => props.backHandler()}
 						/>
 						<ActionBarSeparator />
 						<ActionBarButton
 							align='left'
 							iconId='positron-disconnect-connection'
-							text='Disconnect'
-							disabled={disconnectDisabled}
-							onPressed={() => props.selectedEntry?.disconnect?.()}
+							text={localize('positron.schemaNavigationActionBar.disconnect', 'Disconnect')}
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx:66]: Should be:

```
text={(() => localize('positron.schemaNavigationActionBar.disconnect', 'Disconnect'))()}
```

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx [hunk 9]
```diff
@@ -73,38 +55,24 @@ export const ActionBar = (props: React.PropsWithChildren<ConnectionActionBarProp
...
+							onPressed={() => props.disconnectHandler()}
 						/>
 						<ActionBarSeparator />
 						<ActionBarButton
 							align='left'
 							iconId='refresh'
-							onPressed={() => props.selectedEntry?.refresh?.()}
-							disabled={props.selectedEntry === undefined || props.selectedEntry.refresh === undefined || !props.selectedEntry.active}
+							text={localize('positron.schemaNavigationActionBar.refresh', 'Refresh')}
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx:73]: Should be:

```
text={(() => localize('positron.schemaNavigationActionBar.refresh', 'Refresh'))()}
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts [hunk 10]
```diff
@@ -53,54 +50,74 @@ class BaseConnectionsInstance extends Disposable {
 
 export class PositronConnectionsInstance extends BaseConnectionsInstance implements IPositronConnectionInstance {
 
-	readonly onToggleExpandEmitter: Emitter<void> = new Emitter<void>();
-	private readonly onToggleExpand: Event<void> = this.onToggleExpandEmitter.event;
+	private readonly onDidChangeEntriesEmitter = new Emitter<IPositronConnectionEntry[]>();
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts:53]:
```
private readonly onDidChangeEntriesEmitter = this._register(new Emitter<IPositronConnectionEntry[]>());
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts [hunk 11]
```diff
@@ -53,54 +50,74 @@ class BaseConnectionsInstance extends Disposable {
...
+	readonly onDidChangeEntries = this.onDidChangeEntriesEmitter.event;
 
-	readonly onDidChangeDataEmitter: Emitter<void>;
+	private readonly onDidChangeStatusEmitter = new Emitter<boolean>();
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts:56]:
```
private readonly onDidChangeStatusEmitter = this._register(new Emitter<boolean>());
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts [hunk 12]
```diff
@@ -53,54 +50,74 @@ class BaseConnectionsInstance extends Disposable {
...
+	readonly onDidChangeStatus = this.onDidChangeStatusEmitter.event;
+
+	public readonly onToggleExpandEmitter = new Emitter<string>();
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts:59]:
```
public readonly onToggleExpandEmitter = this._register(new Emitter<string>());
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts [hunk 13]
```diff
@@ -191,23 +248,25 @@ export class PositronConnectionsInstance extends BaseConnectionsInstance impleme
...
+		readonly connectionsService: IPositronConnectionsService
 	) {
 		super(metadata);
 	}
 
 	readonly kind: string = 'database';
 	readonly expanded: boolean | undefined = false;
 	readonly active: boolean = false;
+	readonly onDidChangeStatus = Event.None;
+	readonly onToggleExpandEmitter = new Emitter<string>();
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts:261]:
```
readonly onToggleExpandEmitter = this._register(new Emitter<string>());
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts [hunk 14]
```diff
@@ -22,17 +21,13 @@ import { INotificationService, Severity } from 'vs/platform/notification/common/
 
 class PositronConnectionsService extends Disposable implements IPositronConnectionsService {
 
-	private readonly _cache: PositronConnectionsCache;
 	readonly _serviceBrand: undefined;
 
-	private onDidChangeEntriesEmitter = new Emitter<IPositronConnectionEntry[]>;
-	onDidChangeEntries: Event<IPositronConnectionEntry[]> = this.onDidChangeEntriesEmitter.event;
+	private onDidChangeConnectionsEmitter = new Emitter<IPositronConnectionInstance[]>;
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts:26]:
```
private onDidChangeConnectionsEmitter = this._register(new Emitter<IPositronConnectionInstance[]>);
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts [hunk 15]
```diff
@@ -22,17 +21,13 @@ import { INotificationService, Severity } from 'vs/platform/notification/common/
...
+	onDidChangeConnections: Event<IPositronConnectionInstance[]> = this.onDidChangeConnectionsEmitter.event;
 
-	public onDidChangeDataEmitter = new Emitter<void>;
-	private onDidChangeData = this.onDidChangeDataEmitter.event;
-
-	public onDidFocusEmitter = new Emitter<void>;
-	private onDidFocus = this.onDidFocusEmitter.event;
+	public onDidFocusEmitter = new Emitter<string>;
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts:29]:
```
	public onDidFocusEmitter = this._register(new Emitter<string>);
```

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts [hunk 16]
```diff
@@ -169,10 +145,15 @@ class PositronConnectionsService extends Disposable implements IPositronConnecti
 			this.connections.push(instance);
 		}
 
+		instance.onDidChangeStatus(() => {
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts:148]: This event handler must be registered.

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsUtils.ts [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+	/***
+	 * The list of connections entries is flat. Level allows us to find
+	 * how nested an entry is.
+	 */
+	level: number;
+
+	/***
+	 * If the entry is expanded or not. Undefined if the entry is not expandable.
+	 */
+	expanded: boolean | undefined;
```
Comment (softwarenerd) [inline src/vs/workbench/services/positronConnections/browser/positronConnectionsUtils.ts:19]: Can be:

```
	/***
	 * If the entry is expanded or not. Undefined if the entry is not expandable.
	 */
	expanded?: boolean;
```

To match `error?: string` below it.

**Review by softwarenerd — CHANGES_REQUESTED** [review]
Impressive haul!

File: src/vs/workbench/contrib/positronConnections/browser/components/listConnections.tsx [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+	);
+};
+
+
+const ACTION_BAR_PADDING_LEFT = 8;
+const ACTION_BAR_PADDING_RIGHT = 8;
+const ACTION_BAR_HEIGHT = 32;
+
+interface ActionBarProps extends PositronConnectionsServices {
+	deleteConnectionHandler?: () => void;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/listConnections.tsx:173]: Resolved in b770b3e84f89293cbdeca3ea43dd73cdb65dfec2

File: src/vs/workbench/contrib/positronConnections/browser/components/resumeConnectionModalDialog.css [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+
+.connections-resume-connection-modal .footer {
+	grid-area: footer;
+	display: flex;
+}
+
+.connections-resume-connection-modal .footer .button {
+	margin-left: auto;
+	padding-left: 15px;
+	padding-right: 15px;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/resumeConnectionModalDialog.css:46]: Resolved in b4fa6f9c2c2a306962b074a0e497fab44026f082

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css (hunk 3 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css:38]: Resolved in b4fa6f9c2c2a306962b074a0e497fab44026f082

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css (hunk 4 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.css:52]: Resolved in b4fa6f9c2c2a306962b074a0e497fab44026f082

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx (hunk 5 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:68]: Resolved in 6ebfa7f5ed4ca3244d728933dc99a42b9b4f9879

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx (hunk 1 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx]: Resolved in 75d5390dbdbcaf6ec923f296646ad6dab6f034b2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx (hunk 2 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:159]: Resolved with 75d5390dbdbcaf6ec923f296646ad6dab6f034b2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+
+interface PositronConnectionsItemProps {
+	item: IPositronConnectionEntry;
+	style?: any;
+	selected: boolean;
+
+	/**
+	 * What happens when a row is selected?
+	 */
+	onSelectedHandler: () => void;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx]: Resolved in b770b3e84f89293cbdeca3ea43dd73cdb65dfec2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx [hunk 21]
```diff
@@ -0,0 +1,301 @@
...
+
+	/**
+	 * On toggle expand handler
+	 */
+	onToggleExpandHandler(id: string): void;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx]: Resolved in b770b3e84f89293cbdeca3ea43dd73cdb65dfec2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx (hunk 6 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigation.tsx:266]: Resolved in 32a93e41534d7082dc61cc597bac7940027a44a4

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx [hunk 22]
```diff
@@ -36,30 +35,13 @@ interface ActionBarProps {
 }
 
 interface ConnectionActionBarProps extends ActionBarProps {
-	selectedEntry: IPositronConnectionEntry | undefined;
-	clearAllHandler: () => void;
+	disconnectHandler: () => void;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx]: Resolved in b770b3e84f89293cbdeca3ea43dd73cdb65dfec2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx [hunk 23]
```diff
@@ -36,30 +35,13 @@ interface ActionBarProps {
...
+	backHandler: () => void;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx]: Resolved in b770b3e84f89293cbdeca3ea43dd73cdb65dfec2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx [hunk 24]
```diff
@@ -36,30 +35,13 @@ interface ActionBarProps {
...
+	refreshHandler: () => void;
```
[the remark this answers is not in the document]
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx]: Resolved in b770b3e84f89293cbdeca3ea43dd73cdb65dfec2

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx (hunk 7 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx]: Resolved in d28f65b5881b7f2035f17782db70a57dd6cac3f3

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx (hunk 8 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx]: Resolved in d28f65b5881b7f2035f17782db70a57dd6cac3f3

File: src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx (hunk 9 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/components/schemaNavigationActionBar.tsx]: Resolved in d28f65b5881b7f2035f17782db70a57dd6cac3f3

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts (hunk 10 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts (hunk 11 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts (hunk 12 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts (hunk 13 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsInstance.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts (hunk 14 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts (hunk 15 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts (hunk 16 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsService.ts]: Resolved in fe6a0d009ddab8ff0dc5ac6b3cfd3145ec8cdd97

File: src/vs/workbench/services/positronConnections/browser/positronConnectionsUtils.ts (hunk 17 above)
Comment (dfalbel) [reply src/vs/workbench/services/positronConnections/browser/positronConnectionsUtils.ts]: Resolved in 6ff77efbc3d0a61d4c9dab516e8fcb3a672a1b42

**Review by softwarenerd — APPROVED** [review]
LG!

**Pull request merged**